### ISSUE TITLE
How to suppress the C++ standard assignment from using the -A option?

### ISSUE THREAD

**Issue by cponder** [kokkos/kokkos#4297]
I'm building Kokkos 3.4.01 under CentOS 7 using the PGI 21.7 compiler.
I'd had to re-direct the C++ header files to use a GCC 7.4.0 install on the system, since C++ 14 extensions aren't supported by the default 4.8.5 compiler.
It looks like the Kokkos is building OK but all of the regression-tests are failing on the system.
The compiler detection phase has these messages
```
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is PGI 21.7.0
-- Using --c++14;-A for C++14 standard as feature
```
Here is the breakage:
```
[  3%] Building CXX object core/unit_test/CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o
cd /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build/core/unit_test && /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/bin/kokkos_
launch_compiler /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/bin/nvcc_wrapper /glade/u/apps/opt/nvhpc/21.7/Linux_x86_64/21.7/compilers/bin/pgc++ /gl
ade/u/apps/opt/nvhpc/21.7/Linux_x86_64/21.7/compilers/bin/pgc++ -DGTEST_HAS_PTHREAD=0 -DGTEST_HAS_TR1_TUPLE=0 -I/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.
4/distro/tpls/gtest -I/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build/core/unit_test -I/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/
distro/core/unit_test -I/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/core/unit_test/category_files -tp=host -noswitcherror -O2 -gopt --c++14 -A -o C
MakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o -c /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/tpls/gtest/gtest/gtest-all.cc
"/usr/include/bits/socket.h", line 257: error: incomplete type is not allowed
      __extension__ unsigned char __cmsg_data __flexarr; /* Ancillary data.  */
                                  ^

1 error detected in the compilation of "/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/tpls/gtest/gtest/gtest-all.cc".
make[2]: [core/unit_test/CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o] Error 2 (ignored)
[  3%] Linking CXX static library libkokkos_gtest.a
cd /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build/core/unit_test && /glade/u/apps/dav/opt/cmake/3.18.2/bin/cmake -P CMakeFiles/kokkos_gtest.dir/cmake_clean_target.cmake
cd /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build/core/unit_test && /glade/u/apps/dav/opt/cmake/3.18.2/bin/cmake -E cmake_link_script CMakeFiles/kokkos_gtest.dir/link.txt --verbose=1
/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/bin/kokkos_launch_compiler /glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/distro/bin/nvcc_wrapper /glade/u/apps/opt/nvhpc/21.7/Linux_x86_64/21.7/compilers/bin/pgc++ /bin/ar qc libkokkos_gtest.a CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o
/bin/ar: CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o: No such file or directory
make[2]: [core/unit_test/libkokkos_gtest.a] Error 1 (ignored)
make[2]: Leaving directory `/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build'
[  3%] Built target kokkos_gtest
```
which causes other breakages later
```
make[2]: *** No rule to make target `core/unit_test/libkokkos_gtest.a', needed by `core/unit_test/KokkosCore_StackTraceTestExec'.
make[2]: Target `core/unit_test/CMakeFiles/KokkosCore_StackTraceTestExec.dir/build' not remade because of errors.
make[2]: Leaving directory `/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build'
make[1]: [core/unit_test/CMakeFiles/KokkosCore_StackTraceTestExec.dir/all] Error 2 (ignored)
[  4%] Built target KokkosCore_StackTraceTestExec
```
and
```
make[2]: *** No rule to make target `core/unit_test/libkokkos_gtest.a', needed by `core/unit_test/KokkosCore_UnitTest_OpenMP'.
make[2]: Target `core/unit_test/CMakeFiles/KokkosCore_UnitTest_OpenMP.dir/build' not remade because of errors.
make[2]: Leaving directory `/glade/work/cponder/SHARE/Utils/Kokkos/3.4.01/PGI-21.7_CUDA-11.4/build'
make[1]: [core/unit_test/CMakeFiles/KokkosCore_UnitTest_OpenMP.dir/all] Error 2 (ignored)
[ 16%] Built target KokkosCore_UnitTest_OpenMP
```
and so on. I saw this reference on the web suggesting that the ```-A``` flag is overly strict:
          https://forums.developer.nvidia.com/t/nvc-google-test-fails-to-compile/177067
I don't see this flag in the ```CMakeCache.txt``` settings, is there a way to suppress it?
I didn't have any of these problems on Ubuntu 18.04, and the message there didn't include the ```-A```:
```
-- Using --c++14 for C++14 standard as feature
```

**Comment by cponder** [comment]
I tried setting this in the build
```
-D Kokkos_CXX_STANDARD:=14
```
but it didn't help.

**Comment by crtrott** [comment]
I don't think think that -A is coming from us. It is likely that cmake adds that? I can confirm that that is what is happening for me with nvhpc 21.5 but gtest actually builds, so I don't get that error.

**Comment by crtrott** [comment]
OK I read up on this.  Apparently -A is used to be more strict about the CXX standard as you said, and cmake adds that when you turn off CMAKE_CXX_EXTENSIONS. But nvcc does choke on enabling CMAKE_CXX_EXTENSIONS with any other host compiler, because it enables stuff like float128 and what not. 

So it turns out with CMake and PGI you either get "-A" added or "--gnu_extensions". If you don't want either you are out of luck ...

**Comment by crtrott** [comment]
with 21.5 I needed these changes to allow a setting of CMAKE_CXX_EXTENSIONS=ON with nvcc as compiler:
```
diff --git a/bin/nvcc_wrapper b/bin/nvcc_wrapper
index 4e52e4d..7dd6179 100755
--- a/bin/nvcc_wrapper
+++ b/bin/nvcc_wrapper
@@ -226,7 +226,7 @@ do
     cuda_args="$cuda_args $1"
     ;;
   #Handle more known nvcc args
-  --expt-extended-lambda|--expt-relaxed-constexpr|--Wno-deprecated-gpu-targets|-Wno-deprecated-gpu-targets)
+  --expt-extended-lambda|--expt-relaxed-constexpr|--Wno-deprecated-gpu-targets|-Wno-deprecated-gpu-targets|-allow-unsupported-compiler)
     cuda_args="$cuda_args $1"
     ;;
   #Handle known nvcc args that have an argument
diff --git a/cmake/kokkos_compiler_id.cmake b/cmake/kokkos_compiler_id.cmake
index 2384726..7169691 100644
--- a/cmake/kokkos_compiler_id.cmake
+++ b/cmake/kokkos_compiler_id.cmake
@@ -156,7 +156,7 @@ ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
   IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 9.2.88)
     MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
   ENDIF()
-  SET(CMAKE_CXX_EXTENSIONS OFF CACHE BOOL "Kokkos turns off CXX extensions" FORCE)
+  #  SET(CMAKE_CXX_EXTENSIONS OFF CACHE BOOL "Kokkos turns off CXX extensions" FORCE)
 ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
   IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 3.8.0)
     MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
diff --git a/cmake/kokkos_test_cxx_std.cmake b/cmake/kokkos_test_cxx_std.cmake
index 707fb00..f44a9ae 100644
--- a/cmake/kokkos_test_cxx_std.cmake
+++ b/cmake/kokkos_test_cxx_std.cmake
@@ -13,11 +13,11 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
       GLOBAL_SET(KOKKOS_USE_CXX_EXTENSIONS ON)
     ENDIF()
   ELSEIF(CMAKE_CXX_EXTENSIONS)
-    IF(KOKKOS_DONT_ALLOW_EXTENSIONS)
-      MESSAGE(FATAL_ERROR "The chosen configuration does not support CXX extensions flags: ${KOKKOS_DONT_ALLOW_EXTENSIONS}. Must set CMAKE_CXX_EXTENSIONS=OFF to continue")
-    ELSE()
+          #    IF(KOKKOS_DONT_ALLOW_EXTENSIONS)
+          #      MESSAGE(FATAL_ERROR "The chosen configuration does not support CXX extensions flags: ${KOKKOS_DONT_ALLOW_EXTENSIONS}. Must set CMAKE_CXX_EXTENSIONS=OFF to continue")
+          #    ELSE()
       GLOBAL_SET(KOKKOS_USE_CXX_EXTENSIONS ON)
-    ENDIF()
+      #    ENDIF()
   ELSE()
     #For trilinos, we need to make sure downstream projects
     GLOBAL_SET(KOKKOS_USE_CXX_EXTENSIONS OFF)
@@ -109,13 +109,13 @@ UNSET(CAN_COMPILE_CPP14 CACHE) #make sure CMake always re-runs this
 # that we can only use host compilers for CUDA builds that use those flags.
 # It also means that extensions (gnu++14) can't be turned on for CUDA builds.

-IF(KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
-  IF(NOT DEFINED CMAKE_CXX_EXTENSIONS)
-    SET(CMAKE_CXX_EXTENSIONS OFF)
-  ELSEIF(CMAKE_CXX_EXTENSIONS)
-    MESSAGE(FATAL_ERROR "NVCC doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
-  ENDIF()
-ENDIF()
+#IF(KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
+        #  IF(NOT DEFINED CMAKE_CXX_EXTENSIONS)
+        #    SET(CMAKE_CXX_EXTENSIONS OFF)
+        #  ELSEIF(CMAKE_CXX_EXTENSIONS)
+        #    MESSAGE(FATAL_ERROR "NVCC doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
+        #  ENDIF()
+        #ENDIF()

 IF(KOKKOS_ENABLE_CUDA)
   # ENFORCE that the compiler can compile CUDA code.
```

Maybe we need a way of allowing this for PGI, at least in contrast to using GCC as the host compiler it didn't lead to errors during compilation with the gnu-extensions enabled (looks like PGI doesn't actually enable all the problematic ones).

**Comment by cponder** [comment]
If I use the PGI compiler on Ubuntu 18.04 I don't see the ```-A``` show up. But I'm using CMake 3.19.4 on there.
The ```-A``` shows up on the system with CentOS 7.8.2003 and CMake 3.18.2 on it.
I'll see about putting the newer CMake on this system and see what happens.

**Comment by crtrott** [comment]
I can confirm that that option does not show up with cmake 3.19.1 on my machine, but is there with an older version I used earlier (3.17.1). So I would recommend for now just going with a newer cmake version. Note that generally detection of nvc++ is much better in 3.20 or 3.21 or so. We worked around some of the issues with older cmake versions in our internal stuff, but its much less likely for you to run into trouble with the newer version of cmake.
